### DISCUSSION TITLE
Extend Voice Commands to execute named scripts

### DISCUSSION THREAD

**Discussion by sT0v** [gopro/labs#1189] [Feature Requests]
Happy Thanksgiving!
This is so much fun, which is why I'm here but I hope you enjoyed your holiday and ignore this overly enthused idea at least for today.

I like how you can define scripts and load/execute them in this way.
- Define - `!SAVEname=script`
- Load - `!Lname`

The QR code execution is amazing, but if the above is possible, could Voice Commands potentially be extended to include these scripts? Essentially split a camel cased script name and expose it to Voice Command?

**Comment by dnewman-gpsw** [comment]
While a good idea, it is unfortunately not practical. All the vocal commands are trained together, so adding even a single command involves retraining all the commands. This ends up beyond the scope of Labs.

**Reply by sT0v** [reply]
Ok makes sense. 

In a similar vein, how can I assign a script to a fourCC macro code and persist it?

Define
`*LVAL="=Ii=Ss=I/100=S/I=E6.25=E*S=E#2"`

Execute
`^LVAL"Current LV $E"`


Is this an alternative syntax to `!SAVEname=script` and `!Lname`? 

I'd like these scripts to remain after rebooting.

**Reply by dnewman-gpsw** [reply]
The use of * makes it permanent.  One system saves the macro to NVM, the other to the SD card.

**Comment by sT0v** [comment]
Will `SHEL=1` be available on GoPro 11?

**Reply by dnewman-gpsw** [reply]
I'm not sure yet, while another Labs update the for HERO11 is likely, not all Labs features will port backwards. Command and control enhancements tend to go backwards, but some of the extensions use core features only on newer cameras. 

The shell is more experimental than most Labs features, I have no idea if there are practical uses.  I've asked for feedback, but none so far.

**Reply by sT0v** [reply]
I am think maybe it could work as follows:
1. Use native Voice Command to record 2-3sec video (this video would contain the custom audio command
2. Download the latest file via python `OpenGoPro`
3. Extract audio and create transcript using local
[GreedyCTCDecoder](https://pytorch.org/audio/stable/tutorials/speech_recognition_pipeline_tutorial.html#generating-transcripts) which will be cleaned up to match `!Lscript` command
4. Somehow send the `!Lscript` `GoProLabs` command to camera via `OpenGoPro`

So if (4) is possible, then I think I could piece together the rest.

**Reply by dnewman-gpsw** [reply]
Open GoPro doesn't currently support Labs commands, which is the reason the $SHEL=1 command was added. There is a good reason for lack of connectivity of Labs and Open GoPro, they literally run on different operating systems. Open GoPro run on Linux and is network aware, Labs runs on the lower level RTOS and is not network aware. Plus these development teams are separate. However, this can be addressed, and the feature has been requested from the Open GoPro team.

**Reply by sT0v** [reply]
Ok makes sense. Thank you sir. Is there a link to the issue/feature request?

**Reply by dnewman-gpsw** [reply]
I've asked them directly, internally.

**Reply by sT0v** [reply]
Ok great, much appreciated sir!
